### PR TITLE
fix plateau scheduler stepping on incomplete epoch

### LINKED ISSUE #8839: `MisconfigurationException` with `ReduceLROnPlateau` when taking a single step
## 🐛 Bug

When using a learning rate scheduler that requires a monitor, such as `ReduceLROnPlateau`, and we set up a `Trainer` to only train a single step with `max_steps=1`, we get a `MisconfigurationException` that says the only monitors we can use in the LR scheduler are the ones associated with the training metrics, _not_ those for validation or testing epochs. This change is reltaively new, and seems to be added in the 1.4.0 release around these lines: 

https://github.com/PyTorchLightning/pytorch-lightning/blob/a64cc373946a755ce6c3aef57c1be607dfe29a0c/pytorch_lightning/trainer/connectors/optimizer_connector.py#L61-L74

### To Reproduce

Interactive notebook to see this bug: https://colab.research.google.com/drive/1gV7gRr9KzVE_oZrGJcX9o2yYngtobNJJ?usp=sharing 

A quick example without the notebook is here: 

```python
import torch
from pytorch_lightning import LightningModule
from torch.utils.data import Dataset

class BoringModel(LightningModule):

    def __init__(self, monitor: str):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

        self.monitor = monitor

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        avg_loss = torch.stack([x["loss"] for x in outputs]).mean()
        self.log(name='train_loss_epoch', value=avg_loss)

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"x": loss}

    def validation_epoch_end(self, outputs) -> None:
        avg_loss = torch.stack([x['x'] for x in outputs]).mean()
        self.log(name='val_loss_epoch', value=avg_loss)

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler_dict = {
            'scheduler': torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer),
            'monitor': self.monitor,
        }
        return [optimizer], [lr_scheduler_dict]


def test_x(tmpdir, monitor: str):
    # init model
    model = BoringModel(monitor=monitor)

    # Initialize a trainer
    trainer = pl.Trainer(
        max_steps=1,  # note that only a single step is taken here
    )

    # Train the model
    trainer.fit(model, train, val)
```

This works without error: 

```python
>>> test_x(tmpdir, monitor='train_loss_epoch')
```

But when we run this code, we get the error: 

```python
>>> test_x(tmpdir, monitor='val_loss_epoch')
MisconfigurationException: ReduceLROnPlateau conditioned on metric val_loss_epoch which is not available. Available metrics are: ['train_loss_epoch']. Condition can be set using `monitor` key in lr scheduler dict
```

### Expected behavior

Expected behavior would be that **the learning rate scheduler would be able to see all available metrics from validation and testing epochs from the start**, and would be able to use that even if we take a step. In the current setup, debugging and testing a pipeline like this by taking a single step does not work, requring instead that all instances of `max_steps=1` be replaced with `max_epochs=1` to work. 

Ideally, both: 

```python
>>> test_x(tmpdir, monitor='train_loss_epoch')
>>> test_x(tmpdir, monitor='val_loss_epoch')
```

would work without error. 

### Environment

> Please copy and paste the output from our environment collection script:
> https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/requirements/collect_env_details.py
> (For security purposes, please check the contents of the script before running it)
> 
> You can get the script and run it with:
> ```bash
> wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/requirements/collect_env_details.py
> python collect_env_details.py
> ```

^ this script no longer exists btw

- PyTorch Lightning Version (e.g., 1.3.0): 1.4.1 
- PyTorch Version (e.g., 1.8): 1.9.0 
- Python version: 3.8 
- OS (e.g., Linux): Linux 
- CUDA/cuDNN version: N/A 
- GPU models and configuration: N/A 
- How you installed PyTorch (`conda`, `pip`, source): pip 
- If compiling from source, the output of `torch.__config__.show()`: N/A 
- Any other relevant information: N/A 

### Additional context

Great library, thank you for development on this!! 😄

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #8839

Added test FAILS without the fix applied.

This fixes an issue with the training loop trying to step an epoch level scheduler when the epoch is incomplete. This happens in the case when e.g. we specify `Trainer(max_steps=x)` where x < one epoch. This usually does not matter, but is problematic when the scheduler monitors a metric from a validation epoch, which does not actually run if we finish training early.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 1]
```diff
@@ -227,7 +227,8 @@ def on_run_end(self) -> List[List[STEP_OUTPUT]]:
         self.trainer.call_hook("on_epoch_end")
         self.trainer.logger_connector.on_epoch_end()
 
-        self.update_lr_schedulers("epoch", update_plateau_schedulers=True)
+        if self._num_training_batches_reached(self.is_last_batch):
+            self.update_lr_schedulers("epoch", update_plateau_schedulers=True)
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:231]: reviewer: the insertion of this condition here is in symmetry with the case here: 
https://github.com/PyTorchLightning/pytorch-lightning/blob/24f0124ddd478e18b76d64a79f1d7b341a0f8646/pytorch_lightning/loops/epoch/training_epoch_loop.py#L133

File: tests/trainer/optimization/test_optimizers.py [hunk 2]
```diff
@@ -405,6 +405,33 @@ def test_lr_scheduler_strict(tmpdir):
         trainer.fit(model)
 
 
+@pytest.mark.parametrize("complete_epoch", [True, False])
+@mock.patch("torch.optim.lr_scheduler.ReduceLROnPlateau.step")
+def test_lr_scheduler_strict_incomplete_epoch(step_mock, tmpdir, complete_epoch):
```
Comment (awaelchli) [inline tests/trainer/optimization/test_optimizers.py:410]: @carmocca opinion, should I combine this test with the above? note it will be a tradeoff between readability/complexity vs. less code dupe.

Comment (carmocca) [reply tests/trainer/optimization/test_optimizers.py:409]: I would :)

Both together still seem readable to me.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 3]
```diff
@@ -227,7 +227,8 @@ def on_run_end(self) -> List[List[STEP_OUTPUT]]:
         self.trainer.call_hook("on_epoch_end")
         self.trainer.logger_connector.on_epoch_end()
 
-        self.update_lr_schedulers("epoch", update_plateau_schedulers=True)
+        if self._num_training_batches_reached(self.is_last_batch):
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:230]: `is_last_batch` wouldn't be necessary here, right? since it's `on_run_end`.

```suggestion
        if self._num_training_batches_reached():
```

I need to try and replace the saved attribute in favor of progress tracking

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: tests/trainer/optimization/test_optimizers.py (hunk 2 above)
Comment (awaelchli) [reply tests/trainer/optimization/test_optimizers.py]: I combined the tests
